ledger/ledger-api-common/src/main/scala/com/digitalasset/ledger/api/validation/CommandsValidator.scala

```diff
@@ -173,3 +173,49 @@ final class CommandsValidator(ledgerId: LedgerId) {
         Left(missingField("command"))
     }
 }
+
+object CommandsValidator {
+
+  case class Submitters[T](actAs: Set[T], readAs: Set[T])
+
+  def effectiveSubmitters(commands: Option[ProtoCommands]): Submitters[String] = {
+    commands.fold(noSubmitters)(effectiveSubmitters)
+  }
+
+  def effectiveSubmitters(commands: ProtoCommands): Submitters[String] = {
+    val effectiveActAs =
+      if (commands.party.isEmpty)
+        commands.actAs.toSet
+      else
+        commands.actAs.toSet + commands.party
+    val effectiveReadAs = commands.readAs.toSet -- effectiveActAs
+    Submitters(effectiveActAs, effectiveReadAs)
+  }
+
+  val noSubmitters: Submitters[String] = Submitters(Set.empty, Set.empty)
+
+  def validateSubmitters(
+      commands: ProtoCommands): Either[StatusRuntimeException, Submitters[Ref.Party]] = {
+    def actAsMustNotBeEmpty(effectiveActAs: Set[Ref.Party]) =
+      if (effectiveActAs.isEmpty)
+        Left(missingField("party or act_as"))
+      else
+        Right(())
```

Comment:
If it's not a one-line if-else-then, I'd suggest adding curly braces.
```suggestion
      if (effectiveActAs.isEmpty) {
        Left(missingField("party or act_as"))
      } else {
        Right(())
      }
```

Reply:
```suggestion
      Either.cond(effectiveActAs.isEmpty, (), missingField("party or act_as"))
```

As suggested to me by @remyhaemmerle-da .

Reply:
Thanks. I knew there's some "syntactic" sugar for this in scalaz, but didn't think of `Either.cond` 👍

Reply:
And in classic scala fashion, the argument for `right` must be given first (aka the "left" side of the argument list).

Reply:
Unfortunately you and I have lost the "treat Either as a pure sum type" vs "treat Left as errors" in Scala idiomverse.